File: NodeJs/backend-node/routes/analytics.js
import express from "express";
import { ScrapeJob, ScrapedPost } from "../database.js";
import { authenticateToken } from "../middleware/auth.js";

const router = express.Router();

router.use(authenticateToken);

const STOPWORDS = new Set([
  // English
  "the", "a", "an", "and", "or", "but", "in", "on", "at", "to", "for", "of", "with", "by", "about", 
  "as", "from", "that", "this", "these", "those", "it", "its", "they", "them", "their", "we", "us", 
  "you", "your", "he", "him", "his", "she", "her", "is", "are", "was", "were", "be", "been", "have", 
  "has", "had", "do", "does", "did", "not", "no", "yes", "can", "will", "would", "should", "just", 
  // Vietnamese
  "và", "của", "cho", "để", "có", "trong", "là", "các", "những", "một", "hai", "này", "được", "bị", 
  "ra", "vào", "lên", "xuống", "đi", "đến", "với", "từ", "ở", "tại", "như", "nhiều", "ít", "đang", 
  "đã", "sẽ", "cũng", "đều", "hơn", "khi", "nào", "gì", "ai", "đâu", "thì", "mà", "nhưng"
]);

router.get("/:id/analytics", async (req, res) => {
  const jobId = req.params.id;

  try {
    const job = await ScrapeJob.findById(jobId);
    if (!job) {
      return res.status(404).json({ detail: "Job profile not found" });
    }
    if (job.user_id.toString() !== req.user.id) {
      return res.status(403).json({ detail: "Not authorized to view analytics for this job" });
    }

    const posts = await ScrapedPost.find({ job_id: jobId });
    
    if (posts.length === 0) {
      return res.json({
        total_posts: 0,
        total_comments: 0,
        total_reactions: 0,
        reactions_breakdown: { like: 0, love: 0, haha: 0, wow: 0, sad: 0, angry: 0 },
        engagement_over_time: [],
        top_authors: [],
        word_cloud: []
      });
    }

    let totalComments = 0;
    let totalReactions = 0;
    const reactionsBreakdown = { like: 0, love: 0, haha: 0, wow: 0, sad: 0, angry: 0 };
    
    const authorCounts = {};
    const overTimeData = {};
    const wordsList = [];

    posts.forEach(p => {
      totalComments += p.comments_count;
      
      const reacts = p.reactions_json || {};
      totalReactions += reacts.total || 0;
      
      Object.keys(reactionsBreakdown).forEach(k => {
        reactionsBreakdown[k] += reacts[k] || 0;
      });

      if (p.author_name) {
        authorCounts[p.author_name] = (authorCounts[p.author_name] || 0) + 1;
      }

      if (p.timestamp) {
        const dateStr = p.timestamp.toISOString().substring(0, 10);
        if (!overTimeData[dateStr]) {
          overTimeData[dateStr] = { posts_count: 0, reactions_count: 0, comments_count: 0 };
        }
        overTimeData[dateStr].posts_count += 1;
        overTimeData[dateStr].reactions_count += reacts.total || 0;
        overTimeData[dateStr].comments_count += p.comments_count;
      }

      if (p.text) {
        const words = p.text.toLowerCase()
          .replace(/[^\w\s]/g, "")
          .split(/\s+/)
          .filter(w => w.length > 1 && !STOPWORDS.has(w));
        wordsList.push(...words);
      }
    });

    const engagementOverTime = Object.keys(overTimeData).sort().map(dStr => ({
      date: dStr,
      posts_count: overTimeData[dStr].posts_count,
      reactions_count: overTimeData[dStr].reactions_count,
      comments_count: overTimeData[dStr].comments_count
    }));

    const topAuthors = Object.entries(authorCounts)
      .map(([name, count]) => ({ author_name: name, posts_count: count }))
      .sort((a, b) => b.posts_count - a.posts_count)
      .slice(0, 10);

    const wordCounts = {};
    wordsList.forEach(w => {
      wordCounts[w] = (wordCounts[w] || 0) + 1;
    });

    const wordCloud = Object.entries(wordCounts)
      .map(([text, value]) => ({ text, value }))
      .sort((a, b) => b.value - a.value)
      .slice(0, 50);

    res.json({
      total_posts: posts.length,
      total_comments: totalComments,
      total_reactions: totalReactions,
      reactions_breakdown: reactionsBreakdown,
      engagement_over_time: engagementOverTime,
      top_authors: topAuthors,
      word_cloud: wordCloud
    });
  } catch (err) {
    console.error("Analytics error in Node.js:", err);
    res.status(500).json({ detail: "Internal server error" });
  }
});

export default router;

File: apps/backend-python/app/routers/analytics.py
from fastapi import APIRouter, Depends, HTTPException
from collections import Counter
import re
from app import crud, database, schemas
from app.routers.auth import get_current_user

router = APIRouter(prefix="/api/jobs", tags=["analytics"])

STOPWORDS = {
    # English
    "the", "a", "an", "and", "or", "but", "in", "on", "at", "to", "for", "of", "with", "by", "about", 
    "as", "from", "that", "this", "these", "those", "it", "its", "they", "them", "their", "we", "us", 
    "you", "your", "he", "him", "his", "she", "her", "is", "are", "was", "were", "be", "been", "have", 
    "has", "had", "do", "does", "did", "not", "no", "yes", "can", "will", "would", "should", "just", 
    # Vietnamese
    "và", "của", "cho", "để", "có", "trong", "là", "các", "những", "một", "hai", "này", "được", "bị", 
    "ra", "vào", "lên", "xuống", "đi", "đến", "với", "từ", "ở", "tại", "như", "nhiều", "ít", "đang", 
    "đã", "sẽ", "cũng", "đều", "hơn", "khi", "như", "nào", "gì", "ai", "đâu", "nào", "thì", "mà", "nhưng"
}

@router.get("/{job_id}/analytics", response_model=schemas.JobAnalyticsResponse)
async def get_job_analytics(
    job_id: str,
    db=Depends(database.get_db),
    current_user=Depends(get_current_user)
):
    db_job = await crud.get_job(db, job_id)
    if not db_job:
        raise HTTPException(status_code=404, detail="Job not found")
    if db_job.user_id != str(current_user.id):
        raise HTTPException(status_code=403, detail="Not authorized to view analytics for this job")
        
    posts = await crud.get_all_posts_by_job(db, job_id)
    if not posts:
        return {
            "total_posts": 0,
            "total_comments": 0,
            "total_reactions": 0,
            "reactions_breakdown": {"like": 0, "love": 0, "haha": 0, "wow": 0, "sad": 0, "angry": 0},
            "engagement_over_time": [],
            "top_authors": [],
            "word_cloud": []
        }
        
    total_posts = len(posts)
    total_comments = sum(p.comments_count for p in posts)
    
    reactions_breakdown = {"like": 0, "love": 0, "haha": 0, "wow": 0, "sad": 0, "angry": 0}
    total_reactions = 0
    
    # Author aggregation
    author_counts = Counter()
    
    # Over time aggregation
    # Key: "YYYY-MM-DD", Value: {posts_count, reactions_count, comments_count}
    over_time_data = {}
    
    # Word frequency aggregation
    words_list = []
    
    for p in posts:
        # Accumulate reactions
        reacts = p.reactions_json or {}
        p_reactions_total = reacts.get("total", 0)
        total_reactions += p_reactions_total
        
        for k in reactions_breakdown.keys():
            reactions_breakdown[k] += reacts.get(k, 0)
            
        # Author counts
        if p.author_name:
            author_counts[p.author_name] += 1
            
        # Over time group
        if p.timestamp:
            date_str = p.timestamp.strftime("%Y-%m-%d")
            if date_str not in over_time_data:
                over_time_data[date_str] = {"posts_count": 0, "reactions_count": 0, "comments_count": 0}
            over_time_data[date_str]["posts_count"] += 1
            over_time_data[date_str]["reactions_count"] += p_reactions_total
            over_time_data[date_str]["comments_count"] += p.comments_count
            
        # Extract words from text
        if p.text:
            # Simple word cleaning: lowercase and strip punctuation
            clean_text = re.sub(r"[^\w\s]", "", p.text.lower())
            words = [w for w in clean_text.split() if w and len(w) > 1 and w not in STOPWORDS]
            words_list.extend(words)

    # Sort over time engagement by date
    engagement_over_time = []
    for d_str, val in sorted(over_time_data.items()):
        engagement_over_time.append(
            schemas.EngagementOverTime(
                date=d_str,
                posts_count=val["posts_count"],
                reactions_count=val["reactions_count"],
                comments_count=val["comments_count"]
            )
        )
        
    # Top authors
    top_authors = [
        schemas.TopAuthor(author_name=name, posts_count=cnt)
        for name, cnt in author_counts.most_common(10)
    ]
    
    # Word cloud
    word_counts = Counter(words_list)
    word_cloud = [
        schemas.WordFreq(text=word, value=cnt)
        for word, cnt in word_counts.most_common(50)
    ]
    
    return schemas.JobAnalyticsResponse(
        total_posts=total_posts,
        total_comments=total_comments,
        total_reactions=total_reactions,
        reactions_breakdown=reactions_breakdown,
        engagement_over_time=engagement_over_time,
        top_authors=top_authors,
        word_cloud=word_cloud
    )
